File: src/main/java/com/refinedmods/refinedstorage/container/factory/PositionalTileContainerProvider.java
package com.refinedmods.refinedstorage.container.factory;

import net.minecraft.core.BlockPos;
import net.minecraft.network.chat.Component;
import net.minecraft.world.MenuProvider;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.level.block.entity.BlockEntity;

import javax.annotation.Nullable;

public class PositionalTileContainerProvider<T extends BlockEntity> implements MenuProvider {
    private final Component name;
    private final Provider<T> provider;
    private final BlockPos pos;

    public PositionalTileContainerProvider(Component name, Provider<T> provider, BlockPos pos) {
        this.name = name;
        this.provider = provider;
        this.pos = pos;
    }

    @Override
    public Component getDisplayName() {
        return name;
    }

    @Nullable
    @Override
    public AbstractContainerMenu createMenu(int windowId, Inventory inventory, Player player) {
        T tile = (T) player.level.getBlockEntity(pos);

        return provider.create(tile, windowId, inventory, player);
    }

    public interface Provider<T> {
        AbstractContainerMenu create(T tile, int windowId, Inventory inventory, Player player);
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/container/factory/PositionalTileContainerProvider.kt
package com.refinedmods.refinedstorage.container.factory

import com.refinedmods.refinedstorage.api.network.grid.IGrid.slotId
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.inventory.container.Container
import net.minecraft.inventory.container.INamedContainerProvider
import net.minecraft.tileentity.BlockEntity
import net.minecraft.util.math.BlockPos
import net.minecraft.util.text.Text

class PositionalTileContainerProvider<T : BlockEntity?>(name: Text, provider: Provider<T>, pos: BlockPos) : INamedContainerProvider {
    interface Provider<T> {
        fun create(tile: T, windowId: Int, inventory: PlayerInventory?, player: PlayerEntity?): Container
    }

    private val name: Text
    private val provider: Provider<T?>
    private val pos: BlockPos
    val displayName: Text
        get() = name

    @Nullable
    fun createMenu(windowId: Int, inventory: PlayerInventory?, player: PlayerEntity): Container {
        val tile = player.world.getBlockEntity(pos) as T?
        return provider.create(tile, windowId, inventory, player)
    }

    init {
        this.name = name
        this.provider = provider
        this.pos = pos
    }
}